src/tqec/sketchup/block_graph.py

```diff
@@ -562,3 +584,45 @@ def display(
             filepath_or_bytes=bytes_buffer.getvalue(),
             write_html_filepath=write_html_filepath,
         )
+
+    def get_abstract_observables(self) -> list[AbstractObservable]:
+        """Get all the abstract observables from the block graph."""
+        self.check_validity(allow_virtual_node=False)
+        correlation_subgraphs = self.to_zx_graph().find_correlation_subgraphs()
+        abstract_observables: list[AbstractObservable] = []
+
+        def is_measured(cube: Cube) -> bool:
+            return self.get_pipe(cube.position, cube.position.shift_by(0, 0, 1)) is None
+
+        for g in correlation_subgraphs:
+            if g.num_nodes == 1:
+                cube = ty.cast(Cube, self.get_cube(g.nodes[0].position))
+                abstract_observables.append(
+                    AbstractObservable(frozenset({cube}), frozenset())
+                )
+                continue
+            top_lines: set[Cube | Pipe] = set()
+            bottom_regions: set[Pipe] = set()
+            for edge in g.edges:
+                pipe = ty.cast(Pipe, self.get_pipe(edge.u.position, edge.v.position))
+                u, v = pipe.u, pipe.v
+                if pipe.direction == Direction3D.Z:
+                    if is_measured(v):
+                        top_lines.add(v)
+                    continue
+                correlation_type_at_src = edge.u.node_type.value
+                surface_attach_direction = Direction3D.from_axis_index(
+                    pipe.pipe_type.value.index(correlation_type_at_src)
+                )
+                if surface_attach_direction == Direction3D.Z:
+                    top_lines.add(pipe)
+                    if is_measured(u):
+                        top_lines.add(u)
+                    if is_measured(v):
+                        top_lines.add(v)
+                else:
+                    bottom_regions.add(pipe)
```

Comment:
I do not understand this part. If `surface_attach_direction` is `Direction.Z` that means that we have CUBE1 -> PIPE -> CUBE2 that are all on the same (x, y) coordinates but with increasing `z` right? In which case, why should the pipe be added?

Reply:
No, here `surface_attach_direction` is `Direction.Z` means that the normal direction of the correlation surface is in the XY plane. Sorry for  the confusing name and I will try to improve that.